+npxtoolkit/+internal/+logging/log4m.m

```diff
@@ -0,0 +1,270 @@
+    end
+    
+    properties(SetAccess = protected)
+        fullpath = 'log4m.log';  %Default file
+        commandWindowLevel = log4m.ALL;
+        logLevel = log4m.INFO;
+    end
+    
+    methods (Static)
+        function obj = getLogger( logPath )
+            %GETLOGGER Returns instance unique logger object.
+            %   PARAMS:
+            %       logPath - Relative or absolute path to desired logfile.
+            %   OUTPUT:
+            %       obj - Reference to signular logger object.
+            %
+            
+            if(nargin == 0)
+                logPath = 'log4m.log';
+            elseif(nargin > 1)
+                error('getLogger only accepts one parameter input');
+            end
+            
+            persistent localObj;
+            if isempty(localObj) || ~isvalid(localObj)
+                localObj = log4m(logPath);
+            end
+            obj = localObj;
+        end
+        
+        function testSpeed( logPath )
+            %TESTSPEED Gives a brief idea of the time required to log.
+            %
+            %   Description: One major concern with logging is the
+            %   performance hit an application takes when heavy logging is
+            %   introduced. This function does a quick speed test to give
+            %   the user an idea of how various types of logging will
+            %   perform on their system.
+            %
+            
+            L = log4m.getLogger(logPath);
+            
+            
+            disp('1e5 logs when logging only to command window');
+            
+            L.setCommandWindowLevel(L.TRACE);
+            L.setLogLevel(L.OFF);
+            tic;
+            for i=1:1e5
+                L.trace('log4mTest','test');
+            end
+            
+            disp('1e5 logs when logging only to command window');
+            toc;
+            
+            disp('1e6 logs when logging is off');
+            
+            L.setCommandWindowLevel(L.OFF);
+            L.setLogLevel(L.OFF);
+            tic;
+            for i=1:1e6
+                L.trace('log4mTest','test');
+            end
+            toc;
+            
+            disp('1e4 logs when logging to file');
+            
+            L.setCommandWindowLevel(L.OFF);
+            L.setLogLevel(L.TRACE);
+            tic;
+            for i=1:1e4
+                L.trace('log4mTest','test');
+            end
+            toc;
+            
+        end
+    end
+    
+    
+%% Public Methods Section %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    methods       
+        function setFilename(self,logPath)
+            %SETFILENAME Change the location of the text log file.
+            %
+            %   PARAMETERS:
+            %       logPath - Name or full path of desired logfile
+            %
+            
+            [fid,message] = fopen(logPath, 'a');
+            
+            if(fid < 0)
+                error(['Problem with supplied logfile path: ' message]);
```

Comment:
Not a great error message. Should say something like "Failed to open log file for writing: ..." and probably include the file path in addition to the error message. "Problem with path" doesn't tell you what kind of problem it encountered, and the `message` that `fopen` returns assumes you already know that it happened in the context of an fopen.